### ISSUE TITLE
Showing also my code here for Legendre polynomials following the Chebyshev example.

### ISSUE THREAD

**Issue by egmaminta** [1ssb/torchkan#3]
Good work! I was wondering if we both have _different_ implementation of Legendre polynomials. This is inspired from the Chebyshev KAN layer.

```
class LegendreKANLayer(nn.Module):
    """ Inspired by Kolmogorov-Arnold Networks but using Legendre polynomials instead of splines coefficients. """
    def __init__(self, input_dim: int, output_dim: int, degree: int):
        super(LegendreKANLayer, self).__init__()
        self.input_dim = input_dim
        self.output_dim = output_dim
        self.degree = degree

        self.legendre_coefficients = nn.Parameter(torch.empty(input_dim, output_dim, degree + 1))
        nn.init.normal_(self.legendre_coefficients, mean=0.0, std=1 / (input_dim * (degree + 1)))

    def forward(self, x):
        x = torch.reshape(x, (-1, self.input_dim)) # shape = (batch_size, input_dim)
        # Since the Legendre polynomial is defined in [-1, 1]
        # We need to normalize x to [-1, 1] using tanh
        x = torch.tanh(x)
        # Initialize the Legendre polynomial tensors
        leg = torch.ones(x.shape[0], self.input_dim, self.degree + 1, device=x.device)
        if self.degree > 0:
            leg[:, :, 1] = x # Set the second Legendre polynomial to x
        for i in range(2, self.degree + 1):
            leg[:, :, i] = ((2 * i - 1) * x * leg[:, :, i-1].clone() - (i - 1) * leg[:, :, i - 2].clone()) / i
        # Compute the Legendre interpolation
        y = torch.einsum('bid,iod->bo', leg, self.legendre_coefficients) # shape = (batch_size, output_dim)
        y = y.view(-1, self.output_dim)
        return y
```

Would appreciate any feedback/comment. Thank you!

**Comment by 1ssb** [comment]
While we share the foundational idea of leveraging Legendre polynomials within a Kolmogorov-Arnold inspired framework, our approaches diverge significantly in structure and computational methods. Your implementation focuses on a single-layer model, whereas mine is a complete multi-layer network. The key differences include:

1. **Polynomial Computations**: You compute Legendre polynomials directly within the layer, while I use a cached function to optimize repeated calculations across multiple layers. I recommend you do this as well.
  
2. **Normalization and Activation**: Your model normalizes inputs with `tanh`, introducing a specific inductive bias, and does not apply an explicit activation on the final output. In contrast, my network uses linear normalization based on the dynamic range of inputs and integrates an activation function after each layer's combined outputs.

3. **Initialization**: You opt for normal distribution for initializing Legendre coefficients. I utilize the Kaiming initializing for both the base and polynomial weights, aiming for a more uniform start in training dynamics.

4. **Structural Flexibility**: Your single-layer design might have less flexibility compared to my cascaded multi-layer structure, which could potentially adapt better to varying complexities due to the interplay of multiple transformations and normalizations.

All of these claims of mine are hypotheses, I will really appreciate if you could prove any or all of them wrong. Feel free to use the `mnist.py` to test your complete model.

**Comment by egmaminta** [comment]
Right! These insights are helpful. I'm still new to this topic and would really appreciate your suggestions. As for your points (2) and (3):

- **Normalization and Activation**: I'm planning on using a linear normalization instead. I do agree with the bias brought by using `.tanh()`. As for the final output, may I politely ask why an explicit activation on the final output could be helpful, and how is it different in terms of performance (if ever you have tried on using one without activation on the final output)?
- **Initialization**: Thank you for pointing this out as well. Will also try Kaiming initialization.

Once again, thank you!

**Comment by 1ssb** [comment]
Here's a reworded version of your message:

---

I haven't thoroughly examined the entire paper, so I don't profess to be an expert. My approach primarily stems from my basic knowledge of MLPs. The first point is straightforward: for a classification task like MNIST, using normalizations along with an explicit activation function is essential. Without these, you're essentially requiring the network to learn the output data manifold directly. Given the transition from high-dimensional input to the low-dimensional classification task, this would create a significant bottleneck in terms of information flow within the model.

**Comment by 1ssb** [comment]
Hi @egmaminta, I haven't thoroughly examined the entire paper, so I don't profess to be an expert. My approach primarily stems from my basic knowledge of MLPs. The first point is straightforward: for a classification task like MNIST, using normalizations along with an explicit activation function is essential. Without these, you're essentially requiring the network to learn the output data manifold directly. Given the transition from high-dimensional input to the low-dimensional classification task, this would create a significant bottleneck in terms of information flow within the model.

**Comment by egmaminta** [comment]
Right. Thank you!
